lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,207 @@
+            access: The desired access level of the dataset (currently either
+                "private" or "public")
+
+        context: The Dataset record, which contains the root run-data index
+            document ID.
+
+        This payload makes use of the Elasticsearch bulk query syntax, which
+        allows combining multiple queries into a single JSON payload using the
+        "Content-Type: application/x-ndjson" header:
+
+        {"index": "jam-pbench.v6.run-data.2001-05"}\n
+        {"authorization.access": "public"}\n
+
+        We rely on the document ID map built by the indexer to find the many
+        Elasticsearch documents created by Pbench indexing. This avoids the
+        need to use multiple Elasticsearch queries here to determine the set
+        of documents on which to act.
+
+        TODO: If some of the bulk updates fail, we need a mechanism either to
+        RETRY or to ROLL BACK (which would in turn require a RETRY); TBD.
+        """
+        user = json_data["user"]
+        name = json_data["name"]
+        access = json_data["access"]
+
+        self.logger.info(
+            "Update access for dataset {} for user {} to {}, prefix {}",
+            name,
+            user,
+            access,
+            self.prefix,
+        )
+
+        map = json.loads(Metadata.get(dataset=context, key=Metadata.INDEX_MAP).value)
+
+        # Construct an "NDJSON" document for a bulk update of all Elasticsearch
+        # documents associated with the Dataset.
+        #
+        # NDJSON is a series of JSON documents separated by "newlines" (defined
+        # by Elasticsearch as cr-lf pairs). We'll construct stringified JSON in
+        # an array which is then gathered up using str.join().
+        #
+        # TODO: it'd probably be better to use the Elasticsearch bulk helper and
+        # pyesbulk here; but I don't want to duplicate the setup in indexer.py,
+        # it's not well generalized, and I don't want to add that refactoring to
+        # this PR.
+
+        cmds = []
+        for index, ids in map.items():
+            for id in ids:
+                x = (
+                    f'{{"update": {{"_index": "{index}", "_id": "{id}"}}}}\n'
+                    f'{{"doc": {{"authorization": {{"access": "{access}"}}}}}}'
+                )
+                cmds.append(x)
+        data = "\n".join(cmds) + "\n"
+        return {
+            "path": "/_bulk",
+            "kwargs": {
+                "data": data,
+                "headers": {"Content-Type": "application/x-ndjson"},
+                "params": {"refresh": "true"},
+            },
+        }
+
+    def postprocess(
+        self, es_json: Dict[AnyStr, Any], context: Dataset
+    ) -> Dict[AnyStr, Any]:
+        """
+        A bulk update succeeds if any of the individual updates succeeds, and
+        PUBLISH can affect hundreds of documents across many Elasticsearch
+        indices. Here we'll look for the "errors" boolean in the response
```

Comment:
It can involve hundreds of documents from across something like a dozen indices, depending on what tools are in the dataset. Which is why the metadata document maps can get to be hundreds of thousands of characters long and I ended up truncating them in the DB dump.

But for each time-based root index name (run, result, etc.), there's a single index name based on the metadata timestamp in the tarball _not_ based on the current time when the index action is generated.